File: legacy-python/work_agent/classify.py
from __future__ import annotations

from .models import WorkItem

URGENCY_RANK = {"high": 0, "medium": 1, "low": 2}
_HIGH_PRIORITY_NAMES = {"highest", "high"}


def classify(item: WorkItem) -> WorkItem:
    """Deterministic, rule-based classification — no ML/LLM call in Phase 1.

    category one of: needs_review, needs_action, blocked, waiting_on_others,
    slack_mention, fyi.
    urgency one of: high, medium, low.
    """
    category = "fyi"
    urgency = "low"
    reasons: list[str] = []

    for pr in item.prs:
        if pr.review_requested_of_me and pr.state == "open":
            category = "needs_review"
            urgency = "high"
            reasons.append(f"Review requested on {pr.repo}#{pr.number}")
        elif pr.is_author and pr.review_state == "changes_requested":
            category = "needs_action"
            urgency = "high"
            reasons.append(f"Changes requested on your PR {pr.repo}#{pr.number}")
        elif pr.is_author and pr.state == "open" and category == "fyi":
            category = "waiting_on_others"
            urgency = "medium"
            reasons.append(f"Your PR {pr.repo}#{pr.number} is awaiting review")

    if item.jira:
        issue = item.jira
        if issue.status.lower() == "on hold":
            if category == "fyi":
                category = "blocked"
                urgency = "medium"
            reasons.append(f"{issue.key} is On Hold")
        elif issue.status_category == "In Progress" and not item.prs:
            if category == "fyi":
                category = "needs_action"
                urgency = "medium"
            reasons.append(f"{issue.key} is in progress with no linked PR yet")

        if issue.priority.lower() in _HIGH_PRIORITY_NAMES and URGENCY_RANK[urgency] > URGENCY_RANK["high"]:
            urgency = "high"
            reasons.append(f"{issue.key} priority is {issue.priority}")

    if item.slack_messages and category == "fyi":
        category = "slack_mention"
        reasons.append(f"{len(item.slack_messages)} Slack message(s) reference this")

    item.category = category
    item.urgency = urgency
    item.reasons = reasons
    return item


def classify_all(items: list[WorkItem]) -> list[WorkItem]:
    return [classify(i) for i in items]

File: src/pipeline/classify.ts
import type { Category, Urgency, WorkItem } from '../models/types.js';

const URGENCY_RANK: Record<Urgency, number> = { high: 0, medium: 1, low: 2 };
const HIGH_PRIORITY_NAMES = new Set(['highest', 'high']);

/** Deterministic, rule-based classification — no ML/LLM call in Phase 1. */
export function classify(item: WorkItem): WorkItem {
  let category: Category = 'fyi';
  let urgency: Urgency = 'low';
  const reasons: string[] = [];

  for (const pr of item.prs) {
    if (pr.reviewRequestedOfMe && pr.state === 'open') {
      category = 'needs_review';
      urgency = 'high';
      reasons.push(`Review requested on ${pr.repo}#${pr.number}`);
    } else if (pr.isAuthor && pr.reviewState === 'changes_requested') {
      category = 'needs_action';
      urgency = 'high';
      reasons.push(`Changes requested on your PR ${pr.repo}#${pr.number}`);
    } else if (pr.isAuthor && pr.state === 'open' && category === 'fyi') {
      category = 'waiting_on_others';
      urgency = 'medium';
      reasons.push(`Your PR ${pr.repo}#${pr.number} is awaiting review`);
    }
  }

  if (item.jira) {
    const issue = item.jira;
    if (issue.status.toLowerCase() === 'on hold') {
      if (category === 'fyi') {
        category = 'blocked';
        urgency = 'medium';
      }
      reasons.push(`${issue.key} is On Hold`);
    } else if (issue.statusCategory === 'In Progress' && item.prs.length === 0) {
      if (category === 'fyi') {
        category = 'needs_action';
        urgency = 'medium';
      }
      reasons.push(`${issue.key} is in progress with no linked PR yet`);
    } else if (issue.statusCategory === 'To Do' && item.prs.length === 0) {
      // A ticket assigned to you that you haven't started is real, visible
      // open work — not "informational." Landing it in the same low-signal
      // "fyi" bucket as an unrelated PR someone mentioned once hid it behind
      // a collapsed section by default, indistinguishable from actual noise.
      if (category === 'fyi') {
        category = 'needs_action';
      }
      reasons.push(`${issue.key} is assigned to you and hasn't been started yet`);
    }

    if (HIGH_PRIORITY_NAMES.has(issue.priority.toLowerCase()) && URGENCY_RANK[urgency] > URGENCY_RANK.high) {
      urgency = 'high';
      reasons.push(`${issue.key} priority is ${issue.priority}`);
    }
  }

  // A merged PR means the ticket is effectively done from your side — a
  // Slack message asking about progress from before it merged is stale
  // context now, not a live reason to surface this as needing attention.
  // ("Recently merged" already gives the merge itself its own visibility.)
  const hasMergedPr = item.prs.some((pr) => pr.state === 'merged');
  if (item.slackMessages.length > 0 && category === 'fyi' && !hasMergedPr) {
    category = 'slack_mention';
    reasons.push(`${item.slackMessages.length} Slack message(s) reference this`);
  }

  item.category = category;
  item.urgency = urgency;
  item.reasons = reasons;
  return item;
}

export function classifyAll(items: WorkItem[]): WorkItem[] {
  return items.map(classify);
}
